### PR TITLE
Add enum type to rpc registry for consolidating RPC initialization code path

### PR DESCRIPTION
**Pull request by xush6528**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28628 Add enum type to rpc registry for consolidating RPC initialization code path**

Consolidate code paths of ProcessGroupAgent construction and other RPC Backend construction.

Differential Revision: [D5516188](https://our.internmc.facebook.com/intern/diff/D5516188/)

### CODE REVIEW

File: test/dist_autograd_test.py
Comment (mrshenli) [inline test/dist_autograd_test.py]: Let's add a comments here to mention that this is needed because this test case calls `dist.barrier()` in the end.

File: test/rpc_test.py
Comment (mrshenli) [inline test/rpc_test.py]: why removing this test?

File: test/rpc_test.py
Comment (mrshenli) [inline test/rpc_test.py]: Why this test does not require a pg? ProcessGroupAgent would ignore `requires_process_group_agent` and still need the the pg instance to collect names and check if they match, no?

File: torch/distributed/rpc/backend_registry.py
Comment (mrshenli) [inline torch/distributed/rpc/backend_registry.py]: why do we need to have two dicts here instead of directly inserting the new backend?

File: torch/distributed/rpc/backend_registry.py
Comment (mrshenli) [inline torch/distributed/rpc/backend_registry.py]: Is this function just used once in `register_backend`? If that's the case, would it be better to make it inline?

File: test/rpc_test.py
Comment (xush6528) [reply test/rpc_test.py]: Because we are changing to Python Enum class. 

`backend_name` string got tested by Enum class internals, which is a Python standard library thing. No need to test Python std library.

File: test/rpc_test.py
Comment (xush6528) [reply test/rpc_test.py]: Because in this PR, I move the pg initialized inside of `def _init_rpc` instead, `def process_group_init_backend_handler` to be precise.

No, I changed `@requires_process_group_agent` to test agianst test config instead of the current agen instance that has been initialzied.

File: torch/distributed/rpc/backend_registry.py
Comment (xush6528) [reply torch/distributed/rpc/backend_registry.py]: Becuase Enum type is a immutable datastrucuture. There is no API from the std library to insert a new enum variant. I have todeclare a new custom Enum class dynamically.

File: torch/distributed/rpc/backend_registry.py
Comment (xush6528) [reply torch/distributed/rpc/backend_registry.py]: Yes, it's not an external API anymore. `backend_name` validity is tested when accessing a particular invariont of the enum type.

For this short one, it's ok to make it inline.

**Comment by xush6528** [comment]
Address @mrshenli 's comments

File: torch/distributed/rpc/backend_registry.py
Comment (mrshenli) [inline torch/distributed/rpc/backend_registry.py]: This is a side effect of this method. Shall we add docs to `init_model_parallel` to describe this?

File: torch/distributed/rpc/backend_registry.py
Comment (mrshenli) [inline torch/distributed/rpc/backend_registry.py]: Is this the only function that's not available in `torch.distributed` and used in this file? If yes, shall we import this function explicitly from `distributed_c10d`. The other two functions `is_initialized` and `init_process_group` can be called from `torch.distributed`.

File: torch/distributed/rpc/backend_registry.py
Comment (mrshenli) [reply torch/distributed/rpc/backend_registry.py]: I feel a better solution might be letting ProcessGroupAgent to have its own pg instance instead of modifying the global one. But I am OK to land this as is to unblock.

File: torch/distributed/rpc/backend_registry.py
Comment (mrshenli) [reply torch/distributed/rpc/backend_registry.py]: I was talking about have two dicts here. I see.

File: torch/distributed/rpc/__init__.py
Comment (mrshenli) [inline torch/distributed/rpc/__init__.py]: I recall that we ran into an issue of rendezvous twice when using `ProcessGroupAgent` when doing it here. Wouldn't this hit that problem again?

**Review by mrshenli — APPROVED** [review]
Stamp to unblock. Please address test failure and comments before landing.

File: torch/distributed/rpc/backend_registry.py
Comment (xush6528) [reply torch/distributed/rpc/backend_registry.py]: Yes, this will caues confusion to users. We could address this later. 

The work comes from the fact that all helper functions in `distributed_c10d.py` modify the global varaiables like `_default_pg` and `_pg_map`. We need to refactor distributed_c10d.py` a bit to have basic building block that only returns a `pg` withought manipulate the global variables. Then, put a `_model_parallel_pg` in `model_parallel/__init__.py`

File: torch/distributed/rpc/__init__.py
Comment (xush6528) [reply torch/distributed/rpc/__init__.py]: I do handle this by inserting a few process_group barrier to circumvent it. 
Putting barrier between failed rendezvous versions could avoid mis-match.

File: test/dist_autograd_test.py
Comment (xush6528) [reply test/dist_autograd_test.py]: Added.

**Comment by xush6528** [comment]
`test_invalid_names` failed due to the reason as described in #28552.
Adding a temporary hack to circumvent it.

File: test/dist_autograd_test.py
Comment (pietern) [reply test/dist_autograd_test.py]: Too much duplication. Consider adding a `with run_barrier_on_exit()`, or a test decorator, for example.

File: test/dist_autograd_test.py
Comment (pietern) [inline test/dist_autograd_test.py]: Please add an issue to fix this and link it in the skip description.

File: test/dist_utils.py
Comment (pietern) [inline test/dist_utils.py]: Why the rename?

File: test/rpc_test.py
Comment (pietern) [inline test/rpc_test.py]: Will this file be removed from disk?

File: torch/distributed/rpc/__init__.py
Comment (pietern) [reply torch/distributed/rpc/__init__.py]: What do you mean here? Mismatch between torch.distributed and torch.distributed.rpc?

File: torch/distributed/rpc/backend_registry.py
Comment (pietern) [inline torch/distributed/rpc/backend_registry.py]: Not a fan of this approach. It means that `BackendType.FOO_BAR` might fail depending on when you call it.

What's wrong with using a string? Common strings can still be exposed as constant.

**Review by pietern — COMMENTED** [review]
There is a lot to unpack in this PR. I feel it should have been a stack of 3 or 4 PRs instead of one big one.

The title lets me know what was done (the enum type), but not why.

What is the reason for using an enum here?

**Comment by xush6528** [comment]
@pietern 

```py
WorkerCtx.backend() is rpc.backend_registry.BackendType.PROCESS_GROUP
```
The benifit of using enum for this use is it's object equality test instead of string comparison.

File: test/rpc_test.py
Comment (xush6528) [reply test/rpc_test.py]: None of the temp file is removed.

https://github.com/pytorch/pytorch/blob/6eef4690743a31ccc546f0c4f2d057d602e1e79e/test/common_distributed.py#L155

File: torch/distributed/rpc/__init__.py
Comment (xush6528) [reply torch/distributed/rpc/__init__.py]: Mismatch between multiple calls to `rendezvous`, an issue described in #28552.

File: torch/distributed/rpc/backend_registry.py
Comment (xush6528) [reply torch/distributed/rpc/backend_registry.py]: Updated. Only this function is called through `distributed_c10d` module.

File: torch/distributed/rpc/backend_registry.py
Comment (xush6528) [reply torch/distributed/rpc/backend_registry.py]: I intend to fail in that case.
The failure would be a clear and standalized Python built-in enum module "no attribute" error, clearly indicating users that they should register their custom backend type into the type enum first.

File: test/dist_utils.py
Comment (xush6528) [reply test/dist_utils.py]: Not fully intended, I tink I picked #27940 instead of #27940 when resolving conflicts.

File: test/dist_autograd_test.py
Comment (xush6528) [reply test/dist_autograd_test.py]: Added in #28977.

File: test/dist_utils.py
Comment (pietern) [reply test/dist_utils.py]: I see. I don't have a preference, but don't like anything with `old` or `new` in it, because it implies something is either being introduced or being deprecated. And for these decorators it's not needed if you use the functools wrappers.

File: test/rpc_test.py
Comment (rohan-varma) [inline test/rpc_test.py]: Any reason this was removed? @mrshenli added it in https://github.com/pytorch/pytorch/pull/28376, to fix this test's flakiness.

Comment (xush6528) [reply test/rpc_test.py]: @rohan-varma Since the one successful `dist.init_process_group(..)` on line 302 was removed too, there is no need to wait for all workers to be synced.

Comment (mrshenli) [reply test/rpc_test.py]: as RPC is creating a PG internally, it seems we need to call barrier on that internal PG? @xush6528 @rohan-varma

Comment (xush6528) [reply test/rpc_test.py]: @mrshenli We discussed offline,
since "Connection closed by peer" error should not be thrown from `ProcessGroupAgent` constructor,  the problem might be the `destroy_process_group()` call in the python `except Exception:` section in , `process_group_init_backend_handler(..)`, which is called by `_init_rpc_backend(...)`.

**Pull request closed without merging**
